Legacy widget's preview functionality is broken when the page is moved

anton-vlasenko opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
Legacy Widget block has a preview function, however, it uses a relative URL.
`iframe` cannot load the widgets.php file when the code is no longer running from the /wp-admin/ URL.
Fixes #30049

## How has this been tested?
1. Go to Appearance -> Widgets page.
2. Click on the "plus sign" button at the very bottom of the page.
3. Add a `Legacy Widget` block.
4. Select `Meta` widget from the dropdown.
5. Click somewhere on the background of the page (not on the `Meta` widget itself). You should see a preview of the `Meta` widget now.
6. Open your browser's developer tools and find an `<iframe />` element with the `wp-block-legacy-widget__edit-preview-iframe`  class. Check its `src` attribute.

#### Expected result:
The `src` attribute of the `<iframe />` should contain an absolute URL.
Please see the screenshots section for reference.

## Screenshots <!-- if applicable -->
![Screenshot 2021-08-12 at 20 00 39](https://user-images.githubusercontent.com/43744263/129246203-e3734aba-d1fc-4ec1-b8e8-6d381e3d488d.png)
![Screenshot 2021-08-12 at 19 50 44](https://user-images.githubusercontent.com/43744263/129246272-05a8fb75-783c-4dbc-9e9a-593a7cd96783.png)


## Types of changes
Bug fix.

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

anton-vlasenko committed:
1. Add new getWPAdminURL and setWPAdminURL functions to be able to set and get WP Admin URL.
This is needed because we have to have the ability to get the actual WordPress admin URL for some blocks (e.g. legacy widgets block).

anton-vlasenko committed:
Use getWPAdminURL instead of addQueryArgs to generate correct absolute url for widgets.php iframe.
We have to use absolute URLs because in some cases (frontenberg) relative URL is not enough and we have to explitly specifiy path to load widgets.php from.

anton-vlasenko committed:
Use getWPAdminURL  from @wordpress/url becase we removed it from utils/url

anton-vlasenko committed:
Add wpAbsoluteAdminUrl  to the block editor settings.
We need this to know absolute URL of the WP's admin panel. Some components need that value.

anton-vlasenko committed:
Revert the changes as we are not going to implement it like that.

anton-vlasenko committed:
Use an absolute URL for widget preview.
It fixes the issue described in https://github.com/WordPress/gutenberg/issues/30049.

anton-vlasenko committed:
Fix code style (automatic).

anton-vlasenko committed:
Rename the variable.

anton-vlasenko committed:
Fix code style.

anton-vlasenko committed:
Remove redundant line at the end of the file.

kevin940726 commented on lib/editor-settings.php:
```diff
@@ -28,6 +28,9 @@ function gutenberg_extend_post_editor_settings( $settings ) {
 	}
 
+	// Some blocks have to use absolute url.
+	$settings['wpAbsoluteAdminUrl'] = admin_url();
```
Nitpick: I think maybe we should just name it `adminUrl` to simplify it a bit.

anton-vlasenko replied:
I agree. The name you proposed is better.
Fixed in b30ec6bc9acb149caf9ded0d8f615e714eb2964d

kevin940726 commented on packages/widgets/src/blocks/legacy-widget/edit/preview.js:
```diff
@@ -11,9 +11,19 @@ import { addQueryArgs } from '@wordpress/url';
 
 export default function Preview( { idBase, instance, isVisible } ) {
 	const [ isLoaded, setIsLoaded ] = useState( false );
+	const settings = useSelect( ( select ) => {
```
Nitpick: we don't need all the editor settings here, so maybe we can just destruct the admin url.

```suggestion
	const { wpAbsoluteAdminUrl } = useSelect( ( select ) => {
```

anton-vlasenko replied:
Yes, I agree.
Fixed in 36370d9820206909f8821e746d56c2f3af25f8c2

kevin940726 commented on packages/widgets/src/blocks/legacy-widget/edit/preview.js:
```diff
@@ -11,9 +11,19 @@ import { addQueryArgs } from '@wordpress/url';
+		return select( blockEditorStore ).getSettings();
+	}, [] );
+
+	const widgetPreviewUrl = settings.wpAbsoluteAdminUrl + 'widgets.php';
```
Nitpick: If `wpAbsoluteAdminUrl` for some reasons isn't defined then the url will be broken. I'm not sure if there's any case where it will not be defined though. But doesn't hurt to add some guard around it?

anton-vlasenko replied:
Good catch!
Fixed in 36370d9820206909f8821e746d56c2f3af25f8c2

kevin940726 reviewed (commented):
This generally looks pretty good! I'm not sure whether it's the best place to add the editor settings though, could get some 👀 from PHP folks too.

anton-vlasenko committed:
1. Check if the admin_url function exists before calling it.
2. Rename wpAbsoluteAdminUrl to adminUrl. IMO This name is shorter and better.

anton-vlasenko committed:
1. Use destruct assignment to get the value of admin URL.
2. Don't assume adminUrl is defined. If it's not defined, return empty string instead.

anton-vlasenko committed:
1. Use destruct assignment to get the value of admin URL.
2. Don't assume adminUrl is defined. If it's not defined, return empty string instead.

anton-vlasenko committed:
Fix code style.

hellofromtonya commented on lib/editor-settings.php:
```diff
@@ -28,6 +28,9 @@ function gutenberg_extend_post_editor_settings( $settings ) {
 	}
 
+	// Some blocks have to use absolute url.
+	$settings['adminUrl'] = is_callable( 'admin_url' ) ? admin_url() : '';
```
`function_exists` is the default pattern for determining if a function is in memory for use

I'm wondering: Should a REST endpoint be used instead of the backend admin URL?

Gutenberg is using the REST API and is not loading the full backend functionality. I'm wondering if using `admin_url()` here, assuming it's loaded into memory at this point, might expose unintentional leaks. 

Hmm not sure. Raising the point for consideration. cc @azaozz @TimothyBJacobs

anton-vlasenko replied:
I don't mind using `function_exists`.
Fixed in 198191eccffc2d94da2f214bf991e2b33ccab900

> I'm wondering: Should a REST endpoint be used instead of the backend admin URL?

+1 That's a very good question.

TimothyBJacobs replied:
There is a ticket about creating a REST API endpoint for this whole thing, I'm not sure what the expected timing on that is, https://github.com/WordPress/wordpress-develop/pull/1508. CC: @adamziel.

draganescu replied:
Should we wait for this simple fix to land after we have that endpoint?

This fix will be ported into core, 100% that function will be there. It is not a feature of the plugin, it is a feature of how core initializes the editors. We've used this in e.g. `modify_welcome_panel` to modify links in WP Admin with no issues.

So I think we can safely use it as it does not load more resources, it does not tie Gutenberg to core more, and it's a quick fix for other problems of the future: having the admin url available as a setting,

adamziel replied:
I was away for last two weeks but I'm back now. I will fix the test failures on https://github.com/WordPress/wordpress-develop/pull/1508 and then we'll only be an approval away from merging.

anton-vlasenko replied:
@draganescu
> Should we wait for this simple fix to land after we have that endpoint?

Maybe I'm not getting it, but I'd not merge this ticket now because, as @hellofromtonya mentioned, it will expose the backend admin URL. Instead, I'd wait until @adamziel's PR is merged. Then we can implement the fix using a REST endpoint.

adamziel replied:
I fixed the tests as promised, all the reviews are welcome :D

anton-vlasenko committed:
Use function_exists instead of is_callable because using function_exists is the recommended way to do it.

draganescu commented on lib/editor-settings.php:
```diff
@@ -28,6 +28,9 @@ function gutenberg_extend_post_editor_settings( $settings ) {
 	}
 
+	// Some blocks have to use absolute url.
+	$settings['adminUrl'] = function_exists( 'admin_url' ) ? admin_url() : '';
```
Should we better create our own filter? This function is a shim that exists for WordPress 5.7 and may get removed so we should not tie into it. I think it's better to define something like `gutenberg_add_admin_url_to_editor_settings` and run that through `add_filter` just like `gutenberg_extend_post_editor_settings`.

azaozz replied:
Don't think there's a chance `admin_url()` will be missing. It is in [wp-includes/link-template.php](https://github.com/WordPress/wordpress-develop/blob/master/src/wp-includes/link-template.php#L3401) which is [included before the REST API and plugins](https://github.com/WordPress/wordpress-develop/blob/master/src/wp-settings.php#L189).

draganescu commented on packages/widgets/src/blocks/legacy-widget/edit/preview.js:
```diff
@@ -11,9 +11,19 @@ import { addQueryArgs } from '@wordpress/url';
+		return select( blockEditorStore ).getSettings();
+	}, [] );
+
+	const widgetPreviewUrl = ( adminUrl ?? '' ) + 'widgets.php';
```
Will we ever not have `adminUrl`?

draganescu commented:
cc @noisysocks do you think this fix can land like this in 5.8.1 and maybe later we'll just have a nice REST endpoint for rendering legacy widgets?

youknowriad commented on packages/widgets/src/blocks/legacy-widget/edit/preview.js:
```diff
@@ -11,9 +11,19 @@ import { addQueryArgs } from '@wordpress/url';
 
 export default function Preview( { idBase, instance, isVisible } ) {
 	const [ isLoaded, setIsLoaded ] = useState( false );
+	const { adminUrl } = useSelect( ( select ) => {
```
The adminUrl doesn't make sense as a block editor setting. The block-editor package is a generic package that can be used to initialize any kind of block editor, even WP agnostic. That setting seem more suited for the widgets screen settings it seems.

kevin940726 replied:
Hmm, I'm not sure I understand. Does it mean that the `settings` prop in `<BlockEditorProvider>` is only supposed to contain fields defined in [`SETTINGS_DEFAULT`](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/README.md#settings_defaults)? I thought it were just a settings context that users can pass their user-defined fields to every component via React Context. The `block-editor` package doesn't have to be aware of these user-defined fields.

Instead of passing `adminUrl` to every editor from the backend, we can only register it in the widgets editors, but I would still imagine using block editor's `settings` as the canonical way. Is the alternative creating a separate context to pass those values?

youknowriad replied:
The settings passed to the `edit-post` package from the backend are splitted between:

 - Settings kept in the edit-post store because are specific to edit-post
 - Settings passed down to `BlockEditorProvider` because they change the behavior the block editor. 

`BlockEditorProvider` is not meant to be used as a generic settings bag. the simplest solution I can think of is to use React Context for that or a store like edit-post/editor.

At least that's how it is currently.

anton-vlasenko committed:
Use context to get adminUrl value.

anton-vlasenko committed:
Use context to get adminUrl value (customizer page).

noisysocks commented:
> cc @noisysocks do you think this fix can land like this in 5.8.1 and maybe later we'll just have a nice REST endpoint for rendering legacy widgets?

I don't think the issue is that urgent. Happy to wait for a proper fix.

anton-vlasenko committed:
Get rid of the WidgetsSettings and load preview using the new REST API endpoint.

anton-vlasenko commented:
[This](https://github.com/WordPress/gutenberg/pull/34184) is the new PR with the proper fix that leverages the new REST API endpoint.
Let's close this PR when that new PR is merged (just saying).

kevin940726 commented on packages/widgets/src/blocks/legacy-widget/edit/preview.js:
```diff
@@ -69,6 +70,9 @@ export default function Preview( { idBase, instance, isVisible } ) {
 				};
 			}
+			renderPreview( idBase, instance ).then( ( response ) => {
+				iframe.srcdoc = response.preview;
```
It's probably better to set state here instead of manipulating DOM directly. We'll have to create a new state and call the setState function here.

```js
const [ srcDoc, setSrcDoc ] = useState( '' );

renderPreview( idBase, instance ).then( ( response ) => {
  setSrcDoc( response.preview );
} );
```

We'll probably also want to add some error handling and abort the fetch when the component is unmounted too. Which can be a follow-up PR.

kevin940726 commented on packages/widgets/src/blocks/legacy-widget/edit/preview.js:
```diff
@@ -7,24 +7,25 @@ import classnames from 'classnames';
- * Internal dependencies
- */
-import WidgetsSettings from '../../../components/widgets-settings';
+async function renderPreview( idBase, instance ) {
```
Nit: I would probably go for `getPreviewHTML` or something like that but that's just my preference.

anton-vlasenko commented:
Closed in favour of https://github.com/WordPress/gutenberg/pull/34184
